stress-testing/stress-test.py

```diff
@@ -0,0 +1,126 @@
+import asyncio
+import argparse
+import uuid
+import time
+
+from jupyterhub_client.api import JupyterHubAPI, JupyterAPI
+from jupyterhub_client.utils import parse_notebook_cells, tangle_cells
+
+async def run_user_notebook(hub_url, cells, token=None):
+    server_started = False
+    notebook_executed = False
+
+    hub = JupyterHubAPI(hub_url=hub_url, api_token=token)
+
+    try:
+        async with hub:
+            token = await hub.identify_token(hub.api_token)
+            service_name = token['name']
+            username = f'service-{service_name}-{uuid.uuid4()}'
+            user = await hub.ensure_user(username, create_user=True)
+            try:
+                if user['server'] is None:
+                    await hub.create_server(username)
+                    try:
+                        while True:
+                            user = await hub.get_user(username)
+                            if user['server'] and user['pending'] is None:
+                                jupyter = JupyterAPI(hub.hub_url / 'user' / username, hub.api_token)
+                                break
+                            await asyncio.sleep(5)
+                        # Mark servers started successfully
+                        server_started = True
+                        async with jupyter:
+                            kernel_id, kernel = await jupyter.ensure_kernel(kernel_spec=None)
+                            async with kernel:
+                                try:
+                                    for i, (code, expected_result) in enumerate(cells):
+                                        kernel_result = await kernel.send_code(username, code, wait=True)
+                                        if kernel_result != expected_result:
+                                            print(f'Kernel result did not match expected result')
+                                    # Mark notebook executed successfully
+                                    notebook_executed = True
+                                finally:
+                                    print(f"Cleanup: delete kernel of {username}")
+                                    await jupyter.delete_kernel(kernel_id)
+                    finally:
+                        print(f"Cleanup: delete-server of {username}")
+                        await hub.delete_server(username)
+            finally:
+                print(f"Cleanup: make sure no server is being started for user {username}")
+                # Even if hub.create_server(username) doesn't finish, hub may still create the
+                # user's server if the request got to it. Let's make sure there's no server running for the user.
+                # Otherwise, users' single-servers will continue to run and hub will return
+                # "SingleUserNotebookApp mixins:520] Error notifying Hub of activity"
+                resp = await hub.session.delete(hub.api_url / 'users' / username / 'server')
```

Comment:
Can `hub.delete_server` and a `hub.delete_user` be used here?

Reply:
We are actually using `hub.delete_user` at line 62.

The reason why `hub.delete_server` is not used here too is because of a behavior I noticed:

The  `hub.create_server` function returns after it has asked the hub to create the server. But this doesn't mean that the single user server is in a ready state (It could be in a pending state). If task cancellation happens before the single user server is ready, a `hub.delete_server` call will get a a 40X response code (without actually returning this to the caller function) and the single-user server won't get deleted.
So, we could potentially end up with lots of running user servers.